File: Santander_Final_Prediction.py
# -*- coding: utf-8 -*-
"""
Created on Tue Nov 26 16:26:09 2019

@author: kyvenkat
"""

import os
import csv
import numpy as np 
import pandas as pd 
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
import pylab as pl 
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split,cross_val_predict,cross_val_score
from sklearn.metrics import roc_auc_score,confusion_matrix,make_scorer,classification_report,roc_curve,auc
from sklearn.model_selection import StratifiedKFold
from imblearn.over_sampling import SMOTE, RandomOverSampler
from imblearn.under_sampling import ClusterCentroids,NearMiss, RandomUnderSampler
import lightgbm as lgb
import eli5
from eli5.sklearn import PermutationImportance
from xgboost import plot_importance
from sklearn import tree
from xgboost.sklearn import XGBClassifier
import xgboost as xgb
from sklearn import cross_validation, metrics   #Additional scklearn functions
from sklearn.grid_search import GridSearchCV   #Perforing grid search
import graphviz
from pdpbox import pdp, get_dataset, info_plots
from matplotlib import pyplot
import scikitplot as skplt
from scikitplot.metrics import plot_confusion_matrix,plot_precision_recall_curve

##################### Setup Working Directory #################################

os.chdir('C:\\Users\\kyvenkat\\Desktop\\W0376 Backup\\Documents Backup\\Python Files')
os.getcwd()

#################### Load Cabfare_Train Data ##################################

santander = pd.read_csv("C:\\Users\\kyvenkat\\Desktop\\W0376 Backup\\Documents Backup\\train (1).csv")
santander.head(5)
santander.dtypes
############### Converting Required Datatypes #################################

santander['target'] = pd.Categorical(santander['target'])
santander['ID_code'] = santander['ID_code'].astype(object)


################# Missing Values Analysis #####################################

missing_val = pd.DataFrame(santander.isnull().sum())
santander[santander['target'].isnull()].index
missing_val = missing_val.rename(columns = {'index':'variables',0:'Missing Values'})
missing_val_percentage = (missing_val['Missing Values']/len(santander))*100
missing_val.insert(1,"Percentage",missing_val_percentage)
missing_val  =missing_val.sort_values('Missing Values', ascending = False)

##################### Exploratory Data Analysis ###############################
################ Checking the class distribution on 'target' variable#########

target_class = santander['target'].value_counts()
percent_target_class=target_class/len(santander)*100

sns.countplot(santander['target'].values)
plt.title('Distribution of Target Class in  train dataset')

#### distribution of mean values distribution per column 
plt.figure(figsize=(16,8))
santander_attributes=santander.columns.values[1:202]
plt.title('Distribution of mean values per column in train dataset')
sns.distplot(santander[santander_attributes].mean(axis=1),color='blue',kde=True,bins=150,label='train')
###### distribution of skewness values per column 
plt.figure(figsize=(16,8))
sns.distplot(santander[santander_attributes].skew(axis=1),color='green',kde=True,bins=150,label='train')
plt.title('Distribution of skewness values per column in train dataset')

###### distribution of kurtosis values per column 
plt.figure(figsize=(16,8))
sns.distplot(santander[santander_attributes].kurtosis(axis=1),color='red',kde=True,bins=150,label='train')
plt.title('Distribution of Kurtosis values per column in train dataset')

################ Outlier Analysis #############################################

def plot_boxplot(df, features):
    i = 0
    sns.set_style('whitegrid')
    plt.figure()
    fig, ax = plt.subplots(10,5,figsize=(18,24))
    for feature in features:
        i += 1
        plt.subplot(10,5,i)
        sns.boxplot(df[feature])
        plt.xlabel(feature, fontsize=11)
        locs, labels = plt.xticks()
        plt.tick_params(axis = 'x', labelsize = 6, pad =-6)
        plt.tick_params(axis = 'y', labelsize = 6)
plt.show()

features = santander.columns.values[3:10]
plot_boxplot(santander,features)
############ Finding outliers and remove from the dataset######################
def outlier_calculation(x):
    ''' calculating outlier indices and replacing them with NA  '''
    #Extract quartiles
    q75, q25 = np.percentile(santander[x], [75 ,25])
    #Calculate IQR
    iqr = q75 - q25
    #Calculate inner and outer fence
    minimum = q25 - (iqr*1.5)
    maximum = q75 + (iqr*1.5)
    
    santander.loc[santander[x] < minimum,x] = np.nan
    santander.loc[santander[x] > maximum,x] = np.nan 
    
colnames = santander.columns.values[2:202]

for x in colnames:
    outlier_calculation(x)

santander.isnull().sum()

list = []
for x in colnames:
    index_list = santander[santander[x].isnull()].index.tolist()
    list.append(index_list)

list = np.unique(list)

santander = santander.dropna()

######################### Checking correlation ################################

santander_new = santander.drop(['target'], axis =1)
plt.figure(figsize =(30,10))
sns.heatmap(santander_new.corr())
plt.title('Correlation matrix ')

santander_correlations=santander_new[santander_attributes].corr()
santander_correlations=santander_correlations.values.flatten()
santander_correlations=santander_correlations[santander_correlations!=1]
plt.figure(figsize=(16,8))
sns.distplot(santander_correlations, color="Red", label="train")
plt.title('Distribution of correlation values in train dataset')


########################### Split the data ####################################
santander = santander.drop('ID_code', axis =1)

X,y=santander.drop('target',axis=1),santander['target']

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.25,shuffle=True)

###################### Building Models ########################################


### Logistic Regression without class balance

lr_model = LogisticRegression()
lr_model.fit(X_train,y_train)
lr_score=lr_model.score(X_train,y_train)

cv_predict=cross_val_predict(lr_model,X_test,y_test)

cm=confusion_matrix(y_test,cv_predict)

## plotting confusion matrix
ax= plt.subplot()
sns.heatmap(cm, annot=True, ax = ax); #annot=True to annotate cells
ax.set_xlabel('Predicted labels');ax.set_ylabel('True labels'); 
ax.set_title('Confusion Matrix'); 
ax.xaxis.set_ticklabels(['0', '1']); ax.yaxis.set_ticklabels(['0', '1']);
### Finding AUC value 
roc_score=roc_auc_score(y_test,cv_predict)
print('ROC score :',roc_score)

false_positive_rate,recall,thresholds=roc_curve(y_test,cv_predict)
roc_auc=auc(false_positive_rate,recall)
## 1. Accuracy : 0.9162
## 2. FNR : 0.73
## 3. AUC: 0.612

################# Logistic Regression with Class balance using SMOTE ##############################

sm = SMOTE(random_state=42, ratio=1.0)
X_smote,y_smote=sm.fit_sample(X_train,y_train)
X_smote_v,y_smote_v=sm.fit_sample(X_test,y_test)
smote_lr=LogisticRegression()
smote_lr.fit(X_smote,y_smote)
cv_pred=cross_val_predict(smote_lr,X_smote_v,y_smote_v)
cm=confusion_matrix(y_smote_v,cv_pred)

## plotting confusion matrix
ax= plt.subplot()
sns.heatmap(cm, annot=True, ax = ax); #annot=True to annotate cells
ax.set_xlabel('Predicted labels');ax.set_ylabel('True labels'); 
ax.set_title('Confusion Matrix'); 
ax.xaxis.set_ticklabels(['0', '1']); ax.yaxis.set_ticklabels(['0', '1']);

smote_lr_score=smote_lr.score(X_smote,y_smote)
print('Accuracy of the smote_model :',smote_lr_score)

roc_score=roc_auc_score(y_smote_v,cv_pred)
print('ROC score :',roc_score)

false_positive_rate,recall,thresholds=roc_curve(y_smote_v,cv_pred)
roc_auc=auc(false_positive_rate,recall)

## 1. Accuracy :0.79762
## 2. FNR: 0.1885
## 3. AUC: 0.80055



###################### Decision Tree Classifier ##############################

tree_model = DecisionTreeClassifier(max_depth=10,min_samples_split = 500)

tree_model.fit(X_smote,y_smote)

cv_predict=cross_val_predict(tree_model,X_smote_v,y_smote_v)

cm=confusion_matrix(y_smote_v,cv_predict)

## plotting confusion matrix
ax= plt.subplot()
sns.heatmap(cm, annot=True, ax = ax); #annot=True to annotate cells
ax.set_xlabel('Predicted labels');ax.set_ylabel('True labels'); 
ax.set_title('Confusion Matrix'); 
ax.xaxis.set_ticklabels(['0', '1']); ax.yaxis.set_ticklabels(['0', '1']);

tree_model_score=tree_model.score(X_smote,y_smote)
print('Accuracy of the tree_model :',tree_model_score)

false_positive_rate,recall,thresholds=roc_curve(y_smote_v,cv_predict)
roc_auc=auc(false_positive_rate,recall)



## 1. Accuracy :0.6838
## 2. FNR: 0.2753
## 3. AUC: 0.6753

##################### RandomForest Classifier ###############################
forest_model = RandomForestClassifier(max_depth = 90, n_estimators = 500, random_state=0)
forest_model.fit(X_train,y_train)

cv_predict=cross_val_predict(forest_model,X_test,y_test)
cm=confusion_matrix(y_test,cv_predict)

## plotting confusion matrix
ax= plt.subplot()
sns.heatmap(cm, annot=True, ax = ax); #annot=True to annotate cells
ax.set_xlabel('Predicted labels');ax.set_ylabel('True labels'); 
ax.set_title('Confusion Matrix'); 
ax.xaxis.set_ticklabels(['0', '1']); ax.yaxis.set_ticklabels(['0', '1']);

forest_model_score=forest_model.score(X_test,y_test)
print('Accuracy of the forest_model :',forest_model_score)

false_positive_rate,recall,thresholds=roc_curve(y_test,cv_predict)
roc_auc=auc(false_positive_rate,recall)


## 1. Accuracy : 0.90307
## 2. FNR : 0.5211
## 3. AUC : 0.7021


######################## XGBoost Classifier ##################################


train_data = xgb.DMatrix(X_smote, y_smote)
test_data  = xgb.DMatrix(X_smote, y_smote)

watchlist = [(train_data, 'train'), (test_data, 'valid')]


params = {"eta":0.3, "gamma":10, "max_depth":5, 
          "min_child_weight":2,"booster":"gbtree", "subsample":0.5, 'objective': 'binary:logistic'}

xgb_model = xgb.train(param_distributions = params, max_depth=6, n_estimators=100, learning_rate=0.3, gamma=3, objective='binary:logistic',watchlist) 
xgb_model = XGBClassifier(params = params, max_depth=6, n_estimators=100, learning_rate=0.3, gamma=3, objective='binary:logistic')
xgb_model.fit(X_smote,y_smote)

print(xgb_model.feature_importances_)
cv_predict=cross_val_predict(xgb_model,X_smote_v,y_smote_v)
cm=confusion_matrix(y_smote_v,cv_predict)

## plotting confusion matrix
ax= plt.subplot()
sns.heatmap(cm, annot=True, ax = ax); #annot=True to annotate cells
ax.set_xlabel('Predicted labels');ax.set_ylabel('True labels'); 
ax.set_title('Confusion Matrix'); 
ax.xaxis.set_ticklabels(['0', '1']); ax.yaxis.set_ticklabels(['0', '1']);

xgb_model_score=xgb_model.score(X_smote_v,y_smote_v)
print('Accuracy of the xgb_model :',xgb_model_score)

false_positive_rate,recall,thresholds=roc_curve(y_smote_v,cv_predict)
roc_auc=auc(false_positive_rate,recall)

## 1. Accuracy :0.931
## 2. FNR : 0.0681
## 3. AUC :0.9304
############## plotting Important Features ###############################

plot_importance(xgb_model,color='blue',max_num_features=10)
pyplot.show()

##################### Finalizing the model ####################################

---------+-------------------------------------------------+------------+-----------+
|  S.No  |              Model                              |    AUC     |     FNR   |
+--------+-------------------------------------------------+------------+-----------+
|    1   |   Logistic Regression(without class balance)    |   0.612    |   73%     |          
|--------+-------------------------------------------------+------------+-----------+
|    2   |   Logistic Regression(with class balance)       |   0.800    |   18.85%  |      
|--------+-------------------------------------------------+------------+-----------+
|    3   |    Decision Tree Classification Model           |   0.675    |   27.53%  |
|--------+-------------------------------------------------+------------+-----------+
|    4   |    Random Forest Classification Model           |   0.702    |   52.11%  |
|--------+-------------------------------------------------+------------+-----------+
|    5   |    XGBoost Model (Gradient Boosting Method)     |   0.930    |    6.81%  |
 --------+-------------------------------------------------+------------+-----------+

####################### Loading the Test data ################################

santander_test = pd.read_csv("C:\\Users\\kyvenkat\\Desktop\\W0376 Backup\\Documents Backup\\test.csv")
santander_test.head(5)
santander_test.dtypes
santander_test = santander_test.drop(['ID_code'], axis =1)
####################### Predict Customer Transaction on Test Data #############

santander_test_predict = pd.DataFrame(xgb_model.predict(santander_test.values))

santander_test_predict.to_csv("santander_customer_predictions.csv")


plot_importance(xgb_model,color='red',max_num_features=10)
pyplot.show()

File: Santander_Final_Prediction.R
########## Santander Customer Transaction Prediction Problem ##################

### Set Working Directory ####
setwd('C:/Users/kyvenkat/Desktop/W0376 Backup/Documents Backup')
getwd()

#### loading libraries #######################################

x = c("ggplot2", "corrgram", "DMwR", "usdm", "caret", "randomForest", "e1071",
      "DataCombine", "rpart.plot", "rpart",'xgboost','stats','xgboost','SMOTE','ROSE',
      'pROC','ROCR','C50','Matrix')

#load Packages
lapply(x, require, character.only = TRUE)
rm(x)
  
### Loading data ###
santander_train<- read.csv("C:\\Users\\kyvenkat\\Desktop\\W0376 Backup\\Documents Backup\\train (1).csv", header = TRUE) 
santander_test<- read.csv("C:\\Users\\kyvenkat\\Desktop\\W0376 Backup\\Documents Backup\\test.csv", header = TRUE) 

#santander_train<- read.csv("C:\\Users\\Venkatesh K\\Downloads\\train.csv", header = TRUE) 

### Exploratory Data Analysis #########

str(santander_train)

## Changing required data types

santander_train$ID_code<- as.character(santander_train$ID_code)
santander_train$target <- as.factor(santander_train$target)

### Analyzing Missing values 

Missing_Val<-data.frame(sapply(santander_train,function(x) sum(is.na(x)))) 
Missing_Val$Variables<-rownames(Missing_Val)
row.names(Missing_Val)<-NULL
colnames(Missing_Val)<-c("Missing Values","Variables")

#### There is no missing data in the given dataset

##### Outlier Analysis #######################

y<<-NULL
for(i in 3:length(santander_train)){
  z<-santander_train[,i][santander_train[,i]%in% boxplot.stats(santander_train[,i])$out]
  y<-cbind(y,which(santander_train[,i]%in% z))
}
rm(i,z)
santander_out<-apply(y,2,function(x) unique(x))
santander_out<-array(as.numeric(unlist(santander_out)))
santander_out<-unique(santander_out)



# Remove outliers from the data 

santander_train<-santander_train[-c(santander_out),]

### Check the distribution of Dependent Variables

prop.table(table(santander_train['target']))

ggplot(santander_train, aes(target))+theme_classic()+geom_bar(stat = 'count', fill = 'lightblue', show.legend = TRUE)

### Check the summary of Independent variables 

summary(santander_train)

## Check the skewness of Independent variables 

santander_train_new<-santander_train[,-c(1)]

sant_skw<- apply(santander_train_new,2, function(x) skewness(x))

qplot(sant_skw, geom = 'histogram')

### skewness is in between -0.34 to 0.3 which says that variables are more or less symmetric and follow normal distribution

### Dividing train data based on the target variable

santander_train_target0<-santander_train[which(santander_train$target==0),]
santander_train_target1<-santander_train[which(santander_train$target==1),]

plot_density(santander_train_target1[,c(3:50)], ggtheme = theme_classic(),geom_density_args = list(color='red'))


#### Features Correlation

corr_values<-cor(santander_train_new[,-c(1)], method = 'pearson')
corrgram(santander_train_new, order = NULL, lower.panel = panel.shade, upper.panel = panel.pie, text.panel = panel.txt, 
         main = "Correlation Plot")

### which shows there is no correlation between the Independent variables 


############ Outlier Analysis ##############################

y<<-NULL
for(i in 3:202){
  z<-santander_train[,i][santander_train[,i]%in% boxplot.stats(santander_train[,i])$out]
  y<-rbind(y,which(santander_train[,i]%in% z))
}
i=3
length(z)


### Split the data into Train & Test samples
idx<- sample(nrow(santander_train), size = 0.75*nrow(santander_train))
train_data<- santander_train[idx,]
test_data<-santander_train[-idx,]

#### Removing first column which is non significant 
train_data<- train_data[,-c(1)]
test_data<- test_data[,-c(1)]

############ Model Building ###############################

## Logistic Regression without class imbalance

lr_model<-glm(target~.,data = train_data, family = 'binomial')
summary(lr_model)

pred<-predict(lr_model,test_data[,-c(1)], response = 'class')
pred.resp<-ifelse(pred>0.5,1,0)
accuracy <- table(pred.resp, test_data[,"target"])
sum(diag(accuracy))/sum(accuracy)

# 1. Accuracy = 0.9106
# 2. FNR = 0.23415


### Variable Importance
var_imp_lr.model<-data.frame(varImp(lr_model))
var_imp_lr.model$Variables<-row.names(var_imp_lr.model)
row.names(var_imp_lr.model)<-NULL
var_imp_lr.model<-var_imp_lr.model[,c(2,1)]
## AUC curve

prob <- predict(lr_model, newdata=test_data[,-c(1)], type="response")
pred <- prediction(prob, test_data$target)
perf <- performance(pred, measure = "tpr", x.measure = "fpr")
plot(perf, col = 'red')
auc <- performance(pred, measure = "auc")
auc <- auc@y.values[[1]]

# auc: 0.857059



## Important variables in predicting customer transaction in linear regression without class balancing

var_imp_lr.model<-data.frame(varImp(lr_model))
var_imp_lr.model$Variables<-row.names(var_imp_lr.model)
row.names(var_imp_lr.model)<-NULL
var_imp_lr.model<-var_imp_lr.model[,c(2,1)]


prob <- predict(lr_model, newdata=test_data[,-c(1)], type="response")
pred <- prediction(prob, test_data$target)
perf <- performance(pred, measure = "tpr", x.measure = "fpr")
plot(perf, col = 'red')
auc <- performance(pred, measure = "auc")
auc <- auc@y.values[[1]]

########## Logistic Regression with class Balacing 


train.rose <- ROSE(target~., data =train_data,seed=1)$data
lr_model_cb<- glm(target~., data = train.rose, family = 'binomial')

pred<-predict(lr_model_cb,test_data[,-c(1)], response = 'class')
pred.resp<-ifelse(pred>0.5,1,0)
accuracy <- table(pred.resp, test_data[,"target"])
sum(diag(accuracy))/sum(accuracy)

# 1. Accuracy = 0.87964
# 2. FNR = 0.56857

### Variable Importance
var_imp_lr.model_cb<-data.frame(varImp(lr_model_cb))
var_imp_lr.model_cb$Variables<-row.names(var_imp_lr.model_cb)
row.names(var_imp_lr.model_cb)<-NULL
var_imp_lr.model_cb<-var_imp_lr.model_cb[,c(2,1)]
## AUC curve

prob <- predict(lr_model_cb, newdata=test_data[,-c(1)], type="response")
pred <- prediction(prob, test_data$target)
perf <- performance(pred, measure = "tpr", x.measure = "fpr")
plot(perf, col = 'red')
auc <- performance(pred, measure = "auc")
auc <- auc@y.values[[1]]

### auc:0.86132


#### Decision Tree

tree_mod <- C5.0(x = train_data[, -c(1)], y = train_data$target)
summary(tree_mod)
pred<-predict(tree_mod,test_data[,-c(1)])
accuracy <- table(pred, test_data[,"target"])
sum(diag(accuracy))/sum(accuracy)



# 1. Accuracy = 0.87688
# 2. FNR = 0.730297

### Variable Importance
var_imp_tree_mod<-data.frame(varImp(tree_mod))
var_imp_tree_mod$Variables<-row.names(var_imp_tree_mod)
row.names(var_imp_tree_mod)<-NULL
var_imp_tree_mod<-var_imp_tree_mod[,c(2,1)]

## AUC curve

prob <- predict(tree_mod, newdata=test_data[,-c(1)], type = 'prob')[,2]
pred <- prediction(prob, test_data$target)
perf <- performance(pred, measure = "tpr", x.measure = "fpr")
plot(perf, col = 'red')
auc <- performance(pred, measure = "auc")
auc <- auc@y.values[[1]]

# auc: 0.64244


#### Random Forest Classifier 

library(randomForest)

rf_model<-randomForest(target~.,data = train_data,mtry =20, ntree = 10,importance=TRUE)
summary(rf_model)
pred<-predict(rf_model,test_data[,-c(1)])
accuracy <- table(pred, test_data[,"target"])
sum(diag(accuracy))/sum(accuracy)

# 1. Accuracy = 0.89762
# 2. FNR = 0.521126

### Variable Importance
var_imp_rf_model<-data.frame(varImp(rf_model, type=2))
var_imp_rf_model$Variables<-row.names(var_imp_rf_model)
row.names(var_imp_rf_model)<-NULL
var_imp_rf_model<-var_imp_rf_model[,c(2,1)]

## AUC curve

prob <- predict(rf_model, newdata=test_data[,-c(1)], type = 'prob')
pred <- prediction(prob[,2], test_data$target)
perf <- performance(pred, measure = "tpr", x.measure = "fpr")
plot(perf, col = 'red')
auc <- performance(pred, measure = "auc")
auc <- auc@y.values[[1]]
## auc: 0.702122

#### XGBoost Classifier without class imbalance ############################



target = santander_train$target
label = as.integer(santander_train$target)-1

train.label = label[idx]
test.label = label[-idx]

dtrain  <- sparse.model.matrix(target ~ .-1, data=train_data)
dtest   <- sparse.model.matrix(target ~ .-1, data=test_data)
dim(dtrain)
dim(dtest)

param <- list(objective   = "binary:logistic",
              eval_metric = "auc",
              booster = "gbtree",
              max_depth   = 7,
              eta         = 0.1,
              gammma      = 1,
              lambda      = 1,
              colsample_bytree = 0.5,
              min_child_weight = 1)

xgb_model <- xgboost(params  = param,
                     data    = dtrain,
                     label   = train.label, 
                     nrounds = 500,
                     print_every_n = 100,
                     verbose = 1)

pred <- predict(xgb_model, dtest)
pred.resp <- ifelse(pred >= 0.5, 1, 0)

pred_table<-table(pred.resp, test_data[,"target"])
accuracy<-sum(diag(accuracy))/sum(accuracy)

# 1. Accuracy = 0.9174
# 2. FNR = 0.2115

## AUC curve

prob <- predict(xgb_model, newdata=dtest, type = 'prob')
pred <- prediction(prob, test_data$target)
perf <- performance(pred, measure = "tpr", x.measure = "fpr")
plot(perf, col = 'red')
auc <- performance(pred, measure = "auc")
auc <- auc@y.values[[1]]
## auc: 0.88054

mat<-xgb.importance(feature_names = colnames(dtest2),model = xgb_model)
xgb.plot.importance (importance_matrix = mat[1:20])


######################## XG Boost with Class balancing ####################

prop.table(table(santander_train$target)) 

inTrain <- createDataPartition(y = santander_train$target, p = .6, list = F)
train <- santander_train[inTrain,]
testcv <- santander_train[-inTrain,]
inTest <- createDataPartition(y = testcv$target, p = .5, list = F)
test <- testcv[inTest,]
cv <- testcv[-inTest,]
santander_train <- as.factor(santander_train$target)
rm(inTrain, inTest, testcv)

train<-train[,-c(1)]
test<-test[,-c(1)]


i<- grep("target", colnames(train))

train_SMOTE<-SMOTE(target~.,data = train,perc.over = 200, perc.under=100)

table(train_SMOTE$target)

train$target<-as.numeric(levels(train$target))[train$target]
train_SMOTE$target<-as.numeric(levels(train_SMOTE$target))[train_SMOTE$target]


train_smote <- Matrix(as.matrix(train_SMOTE), sparse = TRUE)
test <- Matrix(as.matrix(test), sparse = TRUE)
cv <- Matrix(as.matrix(cv), sparse = TRUE)
train <- Matrix(as.matrix(train), sparse = TRUE)

train_smote_xgb <- xgb.DMatrix(data = train_smote[,-1], label = train_smote[,1])
test_xgb <- xgb.DMatrix(data = test[,-1], label = test[,1])
cv_xgb <- xgb.DMatrix(data = cv[,-1], label = cv[,1])
train_xgb<-xgb.DMatrix(data = train[,-1], label = train[,1])



parameters <- list(
  booster            = "gbtree",          
  silent             = 0,                 
  eta                = 0.3,               
  gamma              = 0,                 
  max_depth          = 6,                 
  min_child_weight   = 1,                 
  subsample          = 1,                 
  colsample_bytree   = 1,                 
  colsample_bylevel  = 1,                 
  lambda             = 1,                 
  alpha              = 0,                 
  objective          = "binary:logistic",   
  eval_metric        = "auc",
  seed               = 1900               
)

watchlist <- list(train  = train_xgb, cv = cv_xgb)


xgb_smote.model <- xgb.train(parameters, train_smote_xgb, nrounds = 500, watchlist)


xgb_smote.predict <- predict(xgb_smote.model, test[,-c(1)])

xgb_smote.predict.prep<-ifelse(xgb_smote.predict >= 0.5,1,0)
accuracy <- table(xgb_smote.predict.prep, test[,1])
sum(diag(accuracy))/sum(accuracy)

## 1. Accuracy = 0.7996
## 2. FNR = 0.718


roc_smote <- roc(test[,1], predict(xgb_smote.model, test[,-c(1)], type = "prob"))
roc_smote$auc

## auc : 0.8125


####################### Saving the Final model######################

RDS.model<- saveRDS(xgb_model, "C:\\Users\\kyvenkat\\Desktop\\W0376 Backup\\Datasets\\Final_XG Boost model_Santander.rds")

##### Loding the saved model #############

Final_model<- readRDS("C:\\Users\\kyvenkat\\Desktop\\W0376 Backup\\Datasets\\Final_XG Boost model_Santander.rds")

######################### Test data #########################

## Loading the test data 


santander_test<- read.csv("C:\\Users\\kyvenkat\\Desktop\\W0376 Backup\\Documents Backup\\test.csv", header = TRUE) 
head(santander_test,5)
santander_test<-santander_test[,-c(1)]
str(santander_test)
################### Finalizing the model and predicting customer transactions on test data #######################

dtest2<-Matrix(as.matrix(santander_test), sparse = TRUE)
dim(dtest2)



### Predict the test data  


xgb = predict(xgb_model,dtest2)
pred <- ifelse(xgb >= 0.5, 1, 0)

### Final test data with customer transaction predictions #############################

Customer_Transaction_predictions<- data.frame(santander_test,"predictions" = pred)

Customer_Transaction_predictions<-Customer_Transaction_predictions[c(201,1:200)]

write.csv(Customer_Transaction_predictions,"Santander_Customer_Transaction_predictions_R.csv",row.names = FALSE)



